thread-manager/src/lib.rs

```diff
@@ -0,0 +1,380 @@
+        }
+    }
+    pub fn set_process_affinity(config: &ThreadManagerConfig) -> anyhow::Result<Vec<usize>> {
+        let chosen_cores_mask = config.default_core_allocation.as_core_mask_vector();
+
+        crate::policy::set_thread_affinity(&chosen_cores_mask);
+        Ok(chosen_cores_mask)
+    }
+
+    pub fn new(config: ThreadManagerConfig) -> anyhow::Result<Self> {
+        let mut core_allocations = HashMap::<ConstString, Vec<usize>>::new();
+        Self::set_process_affinity(&config)?;
+        let mut manager = ThreadManagerInner::default();
+        manager.populate_mappings(&config);
+        for (name, cfg) in config.native_configs.iter() {
+            let nrt = NativeThreadRuntime::new(name.clone(), cfg.clone());
+            manager
+                .native_thread_runtimes
+                .insert(name.clone().into_boxed_str(), nrt);
+        }
+        for (name, cfg) in config.rayon_configs.iter() {
+            let rrt = RayonRuntime::new(name.clone(), cfg.clone())?;
+            manager
+                .rayon_runtimes
+                .insert(name.clone().into_boxed_str(), rrt);
+        }
+
+        for (name, cfg) in config.tokio_configs.iter() {
+            let tokiort = TokioRuntime::new(name.clone(), cfg.clone())?;
+
+            core_allocations.insert(
+                name.clone().into_boxed_str(),
+                cfg.core_allocation.as_core_mask_vector(),
+            );
+            manager
+                .tokio_runtimes
+                .insert(name.clone().into_boxed_str(), tokiort);
+        }
+        Ok(Self {
+            inner: Arc::new(manager),
+        })
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use {crate::ThreadManagerConfig, std::io::Read};
+    #[cfg(target_os = "linux")]
+    use {
+        crate::{CoreAllocation, NativeConfig, RayonConfig, ThreadManager},
+        std::collections::HashMap,
+    };
+
+    #[test]
+    fn configtest() {
+        let experiments = [
+            "examples/core_contention_dedicated_set.toml",
+            "examples/core_contention_contending_set.toml",
+        ];
+
+        for exp in experiments {
+            println!("Loading config {exp}");
+            let mut conffile = std::path::PathBuf::from(env!("CARGO_MANIFEST_DIR"));
+            conffile.push(exp);
+            let mut buf = String::new();
+            std::fs::File::open(conffile)
+                .unwrap()
+                .read_to_string(&mut buf)
+                .unwrap();
+            let cfg: ThreadManagerConfig = toml::from_str(&buf).unwrap();
+            println!("{:?}", cfg);
+        }
+    }
+    // Nobody runs Agave on windows, and on Mac we can not set mask affinity without patching external crate
+    #[cfg(target_os = "linux")]
+    fn validate_affinity(expect_cores: &[usize], error_msg: &str) {
+        let affinity = affinity::get_thread_affinity().unwrap();
+        assert_eq!(affinity, expect_cores, "{}", error_msg);
+    }
+
+    /*  #[test]
+    fn thread_priority() {
```

Comment:
why is this one commented out?

Reply:
This needs root/extra capabilities + realtime kernel features to really be useful. So it is set aside for "future use" for now.

Reply:
might be nice to add that context into the code. We could also just mark this `#[ignore]`

Reply:
Done